ruma-appservice-api/src/lib.rs

```diff
@@ -4,6 +4,69 @@
 
 #![warn(missing_debug_implementations, missing_docs)]
 
+use serde::{Deserialize, Serialize};
+
 pub mod event;
 pub mod query;
 pub mod thirdparty;
+
+/// A namespace defined by an application service.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Namespace {
+    /// A true or false value stating whether this application service has
+    /// exclusive access to events within this namespace.
+    pub exclusive: bool,
+
+    /// A regular expression defining which values this namespace includes.
+    pub regex: String,
+}
+
+/// Namespaces defined by an application service.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Namespaces {
+    /// Events which are sent from certain users.
+    pub users: Vec<Namespace>,
```

Comment:
None of the fields of `Namespaces` are required, so we should use `serde(default)` to have missing ones be interpreted as empty `Vec`s. In serialization, we can then also skip empty ones.

```suggestion
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub users: Vec<Namespace>,
```